lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,254 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "check-do-concurrent.h"
+#include "attr.h"
+#include "scope.h"
+#include "symbol.h"
+#include "../parser/message.h"
+#include "../parser/parse-tree-visitor.h"
+
+namespace Fortran::semantics {
+
+using namespace parser::literals;
+
+// 11.1.7.5 - enforce semantics constraints on a DO CONCURRENT loop body
+class DoConcurrentEnforcement {
+public:
+  DoConcurrentEnforcement(parser::Messages &messages) : messages_{messages} {}
+  std::set<parser::Label> labels() { return labels_; }
+  template<typename T> bool Pre(const T &) { return true; }
+  template<typename T> void Post(const T &) {}
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+  // C1136
+  void Post(const parser::ReturnStmt &) {
+    messages_.Say(currentStatementSourcePosition_,
+        "RETURN not allowed in DO CONCURRENT"_err_en_US);
+  }
+  // C1137
+  void NoImageControl() {
+    messages_.Say(currentStatementSourcePosition_,
+        "image control statement not allowed in DO CONCURRENT"_err_en_US);
+  }
+  void Post(const parser::SyncAllStmt &) { NoImageControl(); }
+  void Post(const parser::SyncImagesStmt &) { NoImageControl(); }
+  void Post(const parser::SyncMemoryStmt &) { NoImageControl(); }
+  void Post(const parser::SyncTeamStmt &) { NoImageControl(); }
+  void Post(const parser::ChangeTeamConstruct &) { NoImageControl(); }
+  void Post(const parser::CriticalConstruct &) { NoImageControl(); }
+  void Post(const parser::EventPostStmt &) { NoImageControl(); }
+  void Post(const parser::EventWaitStmt &) { NoImageControl(); }
+  void Post(const parser::FormTeamStmt &) { NoImageControl(); }
+  void Post(const parser::LockStmt &) { NoImageControl(); }
+  void Post(const parser::UnlockStmt &) { NoImageControl(); }
+  void Post(const parser::StopStmt &) { NoImageControl(); }
+  void Post(const parser::EndProgramStmt &) { NoImageControl(); }
+
+  void Post(const parser::AllocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "ALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US);
+    }
+  }
+  void Post(const parser::DeallocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "DEALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US);
+    }
+    // C1140: deallocation of polymorphic objects
+    if (anyObjectIsPolymorphic()) {
+      messages_.Say(currentStatementSourcePosition_,
```

Comment:
Can we point to the polymorphic objects in what I suppose is a list of objects?

Reply:
This is a placeholder only. It will necessarily change. Specifically, there is nothing to query at this point that says "at this END statement there are a bunch of objects to deallocate and some of them are polymorphic."